### PR TITLE
8328508: Unify the signatures of the methods address_for_index() and index_for() in BOT implementations

### PR DESCRIPTION
**Pull request by lgxbslgx**
Hi all,

This patch unifies the related methods of the block offset table in Serial, Parallel and GC. It mainly removes the methods `SerialBlockOffsetSharedArray::address_for_index`, `SerialBlockOffsetSharedArray::index_for`, `G1BlockOffsetTable::address_for_index` and `G1BlockOffsetTable::index_for` and adds the corresponding methods `addr_for_entry` and `entry_for_addr`. Their usages and the signatures of other methods are also adjusted.

I find some duplicated code of the block offset table in Serial and Parallel (and less in G1). Maybe, we should refactor the block offset table to reduce the duplicated code in the future.

Tier-1 is running locally.

Thanks for taking the time to review.

Best Regards,
-- Guoxiong

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8328508](https://bugs.openjdk.org/browse/JDK-8328508): Unify the signatures of the methods address_for_index() and index_for() in BOT implementations (**Enhancement** - P4)

### Reviewers
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**)
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18405/head:pull/18405` \
`$ git checkout pull/18405`

Update a local copy of the PR: \
`$ git checkout pull/18405` \
`$ git pull https://git.openjdk.org/jdk.git pull/18405/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18405`

View PR using the GUI difftool: \
`$ git pr show -t 18405`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18405.diff">https://git.openjdk.org/jdk/pull/18405.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18405#issuecomment-2010187007)

### CODE REVIEW

File: src/hotspot/share/gc/serial/serialBlockOffsetTable.hpp [hunk 1]
```diff
@@ -46,16 +46,12 @@ class SerialBlockOffsetSharedArray: public CHeapObj<mtGC> {
...
-    void* start_ptr = &_offset_array[start];
-    memset(start_ptr, offset, num_cards);
-  }
+  // Biased array-start of BOT array for fast BOT entry translation
+  uint8_t* _offset_base;
 
-  uint8_t offset_array(size_t index) const {
-    assert(index < _vs.committed_size(), "index out of range");
-    return _offset_array[index];
+  void fill_range(const uint8_t* const start, size_t num_cards, uint8_t offset) {
```
Comment (albertnetymk) [inline src/hotspot/share/gc/serial/serialBlockOffsetTable.hpp:53]: I think the first arg should be `uint8_t* const start`. Then, the `void*` cast below can be dropped.

File: src/hotspot/share/gc/parallel/objectStartArray.inline.hpp [hunk 2]
```diff
@@ -45,21 +45,14 @@ HeapWord* ObjectStartArray::object_start(HeapWord* const addr) const {
...
-
-  uint8_t offset;
-  while (true) {
-    offset = *entry;
-
-    if (offset < CardTable::card_size_in_words()) {
-      break;
-    }
-
+  uint8_t offset = *entry;
```
Comment (albertnetymk) [inline src/hotspot/share/gc/parallel/objectStartArray.inline.hpp:48]: I kind of like the original style, as it has a single de-reference. (It's indeed rather subjective.) Also, this change is completely unrelated to this ticket.

File: src/hotspot/share/gc/serial/serialBlockOffsetTable.hpp (hunk 1 above)
Comment (lgxbslgx) [reply src/hotspot/share/gc/serial/serialBlockOffsetTable.hpp]: Fixed.

File: src/hotspot/share/gc/parallel/objectStartArray.inline.hpp (hunk 2 above)
Comment (lgxbslgx) [reply src/hotspot/share/gc/parallel/objectStartArray.inline.hpp]: Reverted.

Comment (lgxbslgx) [reply src/hotspot/share/gc/parallel/objectStartArray.inline.hpp]: Should I revert the `SerialBlockOffsetTable::block_start_reaching_into_card` too?

**Comment by lgxbslgx** [comment]
> Tier-1 is running locally.

My local laptop crashed just now when running tier1. Its poor performance prevents me from running such too long tests. But the `Pre-submit` tests passed.

**Review by albertnetymk — APPROVED** [review]

**Comment by lgxbslgx** [comment]
@albertnetymk Thanks for your review. Waiting for another review.

**Comment by tschatzl** [comment]
I changed the bug titles in the CRs to not do any kind of markup. Please avoid markup in bug titles. They are directly taken over into git logs, which are primarily text based, and adding markup of the day just messes with tools, and decreases readability.
Thanks.

**Comment by lgxbslgx** [comment]
> I changed the bug titles in the CRs to not do any kind of markup. Please avoid markup in bug titles. They are directly taken over into git logs, which are primarily text based, and adding markup of the day just messes with tools, and decreases readability. Thanks.

Got it. Thanks.

File: src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp [hunk 3]
```diff
@@ -51,9 +51,8 @@ class G1BlockOffsetTable: public CHeapObj<mtGC> {
   // The reserved region covered by the table.
   MemRegion _reserved;
 
-  // Array for keeping offsets for retrieving object start fast given an
-  // address.
-  volatile u_char* _offset_array;  // byte array keeping backwards offsets
+  // Biased array-start of BOT array for fast BOT entry translation
+  volatile u_char* _offset_base;
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp:55]: Maybe the type could be changed to `uint8_t` as well while we are changing this code but feel free to ignore at this point.

**Review by tschatzl — APPROVED** [review]

**Comment by lgxbslgx** [comment]
@albertnetymk @tschatzl Thanks for the review.

/integrate

File: src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp (hunk 3 above)
Comment (lgxbslgx) [reply src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp:55]: Filed https://bugs.openjdk.org/browse/JDK-8329096 to follow up.

**Pull request closed without merging**